### PR TITLE
Add support for multipleQueryParams in annotation-processor

### PR DESCRIPTION
**Pull request by samvaity**
# Description

Update CoreUtils.appendQueryParam to take Object

image

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 1]
```diff
@@ -432,19 +431,36 @@ public static SerializationFormat serializationFormatFromContentType(HttpHeaders
      * @param url the base URL to which the query parameter will be appended.
      * @param key the name of the query parameter (e.g., "api-version", "name", "After").
      * @param value the value of the query parameter
+     * @param delimiter The delimiter to use if the value is a list.
      * @return the updated URL with the appended query parameter.
      */
-    public static String appendQueryParam(String url, String key, String value) {
-        if (value == null) {
+    public static String appendQueryParam(String url, String key, Object value, char delimiter) {
```
Comment (alzimmermsft) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:437]: Instead of changing this existing API, could we create a new API? Ex, `appendMultiQueryParam(String url, String key, List<String> values)`

Or maybe better long-term, have the annotation processor generated code use `UriBuilder` which is more capable of handling this.

File: sdk/clientcore/annotation-processor-test/src/main/java/io/clientcore/annotation/processor/test/implementation/TestInterfaceClientImpl.java [hunk 2]
```diff
@@ -93,7 +93,7 @@ Response<FooListResult> listNextFooListResult(@PathParam(value = "nextLink", enc
                                                       RequestOptions requestOptions);
 
         @HttpRequestInformation(method = HttpMethod.GET, path = "foos", expectedStatusCodes = { 200 })
-        Response<List<Foo>> listFoo(@HostParam("uri") String uri, RequestOptions requestOptions);
+        Response<List<Foo>> listFoo(@HostParam("uri") String uri, @QueryParam(value = "tags", multipleQueryParams = true) List<String> tags, RequestOptions requestOptions);
```
Comment (alzimmermsft) [inline sdk/clientcore/annotation-processor-test/src/main/java/io/clientcore/annotation/processor/test/implementation/TestInterfaceClientImpl.java:96]: Not related to this PR, but is there a PR coming up for when `@QueryParam` has `encoded = true`?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:437]: Also, this API doesn't support whether the value for the query param was already encoded, which we'll need.

File: sdk/clientcore/annotation-processor-test/src/main/java/io/clientcore/annotation/processor/test/implementation/TestInterfaceClientImpl.java (hunk 2 above)
Comment (samvaity) [reply sdk/clientcore/annotation-processor-test/src/main/java/io/clientcore/annotation/processor/test/implementation/TestInterfaceClientImpl.java:96]: yep, next PR. It should add support for encoded path and query params.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java (hunk 1 above)
Comment (samvaity) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java]: yeah, updated the design to have a QueryParameter model so that we can extend it to future needs too of supporting encoded etc.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 3]
```diff
@@ -447,6 +446,39 @@ public static String appendQueryParam(String url, String key, String value) {
...
+        if (url == null || key == null || value == null) {
+            return url;
+        }
+
+        StringBuilder newUrl = new StringBuilder(url);
+        if (!url.contains("?")) {
+            newUrl.append("?");
+        } else if (!url.endsWith("&") && !url.endsWith("?")) {
+            newUrl.append("&");
+        }
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:468]: Creating a new StringBuilder instance and checking for `?` can be expensive in a hot path. We should collect all the query params that should be appended and append them in one in one go. `appendQueryParams(String baseUrl, Map<String, QueryParam> queryParams)`. This method should return the final URL with all query params appended.

File: sdk/clientcore/annotation-processor/src/main/java/io/clientcore/annotation/processor/models/HttpRequestContext.java [hunk 4]
```diff
@@ -384,4 +386,41 @@ public String getParameterName() {
             return parameterName;
         }
     }
+
+    /**
+     * Represents a query parameter.
+     */
+    public static class QueryParameter {
+        private final String value;
+        private final boolean isMultiple;
```
Comment (srnagar) [inline sdk/clientcore/annotation-processor/src/main/java/io/clientcore/annotation/processor/models/HttpRequestContext.java:395]: We should also consider including the delimiter here and default to `,` if it is `null`.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java (hunk 3 above)
Comment (samvaity) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java]: Discussed offline, updated the method usage to Map<String, Object>.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 5]
```diff
@@ -427,23 +429,51 @@ public static SerializationFormat serializationFormatFromContentType(HttpHeaders
...
-     * @return the updated URL with the appended query parameter.
+     * @param host The base URL to which the query parameter will be appended.
+     * @param queryParams A map containing the query parameters and their values.
+     * @param delimiter The delimiter to use if the value is a list.
+     * @return The URL with the appended query parameter.
      */
-    public static String appendQueryParam(String url, String key, String value) {
-        if (value == null) {
-            return url;
+    public static String appendQueryParam(String host, Map<String, Object> queryParams, char delimiter) {
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:439]: nit: call this `appendQueryParams` as we are appending multiple params.

Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:439]: Should we remove `delimiter` and use the default `,`? When we have a usecase to add `,` we can add another overload when we need another delimiter and also, we don't have a way to specify the delimiter in the annotation today.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 6]
```diff
@@ -427,23 +429,51 @@ public static SerializationFormat serializationFormatFromContentType(HttpHeaders
...
+            String key = encodeValue(entry.getKey());
+            Object value = entry.getValue();
+            String valueString;
+
+            if (value instanceof List<?>) {
+                List<?> valueList = (List<?>) value;
+                valueString = valueList.stream()
+                    .map(Object::toString)
+                    .map(CoreUtils::encodeValue)  // Encoding each value
+                    .collect(Collectors.joining(String.valueOf(delimiter)));  // Join with delimiter
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:458]: The delimiter should be encoded too. It might be best to first create the `valueString` and then call `encode`

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 7]
```diff
@@ -427,23 +429,51 @@ public static SerializationFormat serializationFormatFromContentType(HttpHeaders
...
+            } else {
+                valueString = encodeValue(value.toString());
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:460]: If `encode = true` is set on the annotation, we should skip encoding.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java (hunk 6 above)
Comment (samvaity) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java]: Doing encoding in a seperate PR.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java [hunk 8]
```diff
@@ -427,24 +426,42 @@ public static SerializationFormat serializationFormatFromContentType(HttpHeaders
...
-            return url + "&" + key + "=" + value;
-        } else {
-            return url + "?" + key + "=" + value;
+        StringBuilder urlBuilder = new StringBuilder(host);
+        boolean hasExistingQuery = host.contains("?");
+
+        // Process each key-value pair in the queryParams map
+        for (Map.Entry<String, Object> entry : queryParams.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java:446]: We should do a `null` check here to skip over values that are `null`.

File: sdk/clientcore/core/src/test/java/io/clientcore/core/utils/CoreUtilsTests.java [hunk 9]
```diff
@@ -302,7 +317,7 @@ void testAppendNullQueryParam() {
         String key = "name";
         String expected = "https://example.com";
         // Null value for parameter
-        String result = CoreUtils.appendQueryParam(url, key, null);
+        String result = CoreUtils.appendQueryParams(url, null);
```
Comment (srnagar) [inline sdk/clientcore/core/src/test/java/io/clientcore/core/utils/CoreUtilsTests.java:320]: We should also test a non-empty map with one of the keys having a `null` value.

**Review by srnagar — APPROVED** [review]
